auth/src/main/java/com/firebase/ui/auth/ui/phone/PhoneActivity.java

```diff
@@ -14,396 +14,182 @@
-
-    private static final String KEY_VERIFICATION_PHONE = "KEY_VERIFICATION_PHONE";
-    private static final String KEY_STATE = "KEY_STATE";
-
-    private AlertDialog mAlertDialog;
-    @VisibleForTesting CompletableProgressDialog mProgressDialog;
-    private Handler mHandler;
-    private String mPhoneNumber;
-    private String mVerificationId;
-    private Boolean mIsDestroyed = false;
-    private PhoneAuthProvider.ForceResendingToken mForceResendingToken;
-    private VerificationState mVerificationState;
-
-    public static Intent createIntent(Context context, FlowParameters flowParams, Bundle params) {
-        return createBaseIntent(context, PhoneActivity.class, flowParams)
-                .putExtra(ExtraConstants.PARAMS, params);
+    public static Intent createIntent(Context context, FlowParameters params, Bundle args) {
+        return createBaseIntent(context, PhoneActivity.class, params)
+                .putExtra(ExtraConstants.PARAMS, args);
     }
 
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.fui_activity_register_phone);
 
-        mHandler = new Handler();
-        mVerificationState = VerificationState.VERIFICATION_NOT_STARTED;
-        if (savedInstanceState != null && !savedInstanceState.isEmpty()) {
-            mPhoneNumber = savedInstanceState.getString(KEY_VERIFICATION_PHONE);
+        final PhoneProviderResponseHandler handler =
+                ViewModelProviders.of(this).get(PhoneProviderResponseHandler.class);
+        handler.init(getFlowParams());
+        handler.getOperation().observe(this, new ResourceObserver<IdpResponse>(
+                this, R.string.fui_progress_dialog_signing_in) {
+            @Override
+            protected void onSuccess(@NonNull IdpResponse response) {
+                startSaveCredentials(handler.getCurrentUser(), response, null);
+            }
+
+            @Override
+            protected void onFailure(@NonNull Exception e) {
+                handleError(e);
+            }
+        });
 
-            if (savedInstanceState.getSerializable(KEY_STATE) != null) {
-                mVerificationState = (VerificationState) savedInstanceState.getSerializable(KEY_STATE);
+        final PhoneNumberVerificationHandler verificationHandler =
+                ViewModelProviders.of(this).get(PhoneNumberVerificationHandler.class);
+        verificationHandler.init(getFlowParams());
+        verificationHandler.getOperation().observe(this, new ResourceObserver<PhoneAuthCredential>(
+                this, R.string.fui_verifying) {
+            @Override
+            protected void onSuccess(@NonNull PhoneAuthCredential credential) {
+                handler.startSignIn(credential, new IdpResponse.Builder(
+                        new User.Builder(PhoneAuthProvider.PROVIDER_ID, null)
+                                .setPhoneNumber(getPhoneNumber())
```

Comment:
Ok looking at the two methods where we do fragment detection, `getErrorView()` and `getPhoneNumber()`, that's where we both agree we could improve.

Here I see that we use `getPhoneNumber()` in the verification handler to kick off the other handler.

What's the original reason why this flow benefits from being a multi-Fragment activity rather than a series of activities?  Would that help cut down on the patterns we don't like?

Reply:
You actually just gave me a different idea here: I made the phone number pass through the verification view model—not sure why I didn't think of that before. 🤦‍♂️ It still doesn't address the error view thing though, but that's not as bad as the phone number bit IMO.

As for activities, I'm not sure how they would be solving our problem... they would make it harder to talk between them since we'd be relying on a bunch of bundles. Whad'ya think of bfb3b22?

Reply:
Looks better to me!